libspiel/spiel-speaker.c

```diff
@@ -498,14 +498,55 @@ static void
 _setup_pipeline (SpielSpeaker *self, GError **error)
 {
   SpielSpeakerPrivate *priv = spiel_speaker_get_instance_private (self);
-  GstBus *bus;
-  GstElement *convert = gst_element_factory_make ("audioconvert", "convert");
-  GstElement *sink = gst_element_factory_make (
+  g_autoptr (GstBus) bus = NULL;
+  GstElement *convert = NULL;
+  GstElement *sink = NULL;
+
+  convert = gst_element_factory_make ("audioconvert", "convert");
+  if (convert == NULL)
+    {
+      if (error != NULL && *error == NULL)
+        {
+          g_set_error_literal (error, GST_CORE_ERROR, GST_CORE_ERROR_FAILED,
+                               "Failed to create 'convert' element");
+        }
+
+      return;
+    }
+
+  sink = gst_element_factory_make (
       g_getenv ("SPIEL_TEST") ? "fakesink" : "autoaudiosink", "sink");
+  if (sink == NULL)
+    {
+      if (error != NULL && *error == NULL)
+        {
+          g_set_error_literal (error,
+                               GST_CORE_ERROR,
+                               GST_CORE_ERROR_FAILED,
+                               "Failed to create 'autoaudiosink' element; "
+                               "ensure GStreamer Good Plug-ins are installed");
+        }
+
+      gst_object_unref (gst_object_ref_sink (convert));
```

Comment:
This is so weird but seems legit. I don't know why the factory returns floating references.

Reply:
Yeah, it's definitely an odd thing to mix into MT-safe library. I half-expected that to crash, but there it is :)

I either forgot or never knew, but I guess functions can be organized for floating references to avoid big if-then chunks. Oh just in case, I'm pretty sure `g_autoptr (GstObject)` will not know to do a sink+unref.